### PR TITLE
Only remove deleted records form record arrays when saved

### PR DESCRIPTION
**Pull request by bmac**

### CODE REVIEW

File: packages/ember-data/lib/system/record-array-manager.js [hunk 1]
```diff
@@ -57,11 +57,12 @@ export default Ember.Object.extend({
     @method updateRecordArrays
   */
   updateRecordArrays: function() {
-    this.changedRecords.forEach((record) => {
-      if (record.isDeleted()) {
-        this._recordWasDeleted(record);
+    this.changedRecords.forEach((internalModel) => {
+      if (get(internalModel, 'record.isDestroyed') || get(internalModel, 'record.isDestroying') ||
+           (get(internalModel, 'currentState.stateName') === 'root.deleted.saved')) {
```
Comment (ianstarz) [inline packages/ember-data/lib/system/record-array-manager.js]: :+1:

File: packages/ember-data/lib/system/model/states.js [hunk 2]
```diff
@@ -661,6 +661,7 @@ var RootState = {
       // TODO: More robust semantics around save-while-in-flight
       willCommit: Ember.K,
       didCommit: function(internalModel) {
+        internalModel.disconnectRelationships();
```
Comment (igorT) [inline packages/ember-data/lib/system/model/states.js]: Entering the deleted.saved state should handle this most likely

**Comment by bmac** [comment]
@igorT I have updated this pr based on your feedback.

**Comment by bmac** [comment]
@ianstarz Igor explained why its a good idea to remove `disconnectRelationships`/`reconnectRelationships` so I ended up pulling your commit from https://github.com/emberjs/data/pull/3247 into this pr.

**Comment by ianstarz** [comment]
:+1: that's great! I'm glad it could be useful

File: packages/ember-data/lib/system/model/states.js [hunk 3]
```diff
@@ -685,6 +681,7 @@ var RootState = {
       isDirty: false,
 
       setup: function(internalModel) {
+        internalModel.clearRelationships();
```
Comment (igorT) [inline packages/ember-data/lib/system/model/states.js]: We also do this on dematerialize, so it will be called twice now?

File: packages/ember-data/lib/system/relationships/state/has-many.js [hunk 4]
```diff
@@ -156,8 +156,10 @@ ManyRelationship.prototype.findRecords = function() {
   //TODO CLEANUP
   return this.store.findMany(this.manyArray.toArray().map((rec) => rec._internalModel)).
     then(() => {
-      //Goes away after the manyArray refactor
-      this.manyArray.set('isLoaded', true);
+      if (!this.manyArray.get('isDestroyed')) {
```
Comment (igorT) [inline packages/ember-data/lib/system/relationships/state/has-many.js]: This seems scary, why is this needed?

File: packages/ember-data/lib/system/model/states.js (hunk 3 above)
Comment (bmac) [inline packages/ember-data/lib/system/model/states.js]: Should I remove the `internalModel.clearRelationships();` from the `unloadRecord` methods?

File: packages/ember-data/lib/system/relationships/state/has-many.js (hunk 4 above)
Comment (bmac) [inline packages/ember-data/lib/system/relationships/state/has-many.js]: I copied it from this code https://github.com/bmac/data/blob/rebase-3247/packages/ember-data/lib/system/relationships/state/has-many.js#L187-L189. 

I think with the async code its possible to get into a state where the relationship is cleaned up before the `findMany` resolves.

**Comment by bmac** [comment]
ping @igorT

File: packages/ember-data/lib/system/model/states.js (hunk 3 above)
Comment (igorT) [inline packages/ember-data/lib/system/model/states.js]: Hm, I would rely on it there, and not here, but I think we can fix that later when I merge the unload refactor
